agent/engine/docker_task_engine.go

```diff
@@ -1535,7 +1543,33 @@ func (engine *DockerTaskEngine) stopContainer(task *apitask.Task, container *api
 		apiTimeoutStopContainer = engine.cfg.DockerStopTimeout
 	}
 
-	return engine.client.StopContainer(engine.ctx, dockerID, apiTimeoutStopContainer)
+	return engine.stopDockerContainer(dockerID, container.Name, apiTimeoutStopContainer)
+}
+
+// stopDockerContainer attempts to stop the container, retrying only in case of time out errors.
+// If the maximum number of retries is reached, the container is marked as stopped. This is because docker sometimes
+// deadlocks when trying to stop a container but the actual container process is stopped.
+// for more information, see: https://github.com/moby/moby/issues/41587
+func (engine *DockerTaskEngine) stopDockerContainer(dockerID, containerName string, apiTimeoutStopContainer time.Duration) dockerapi.DockerContainerMetadata {
+	var md dockerapi.DockerContainerMetadata
+	backoff := newExponentialBackoff(stopContainerBackoffMin, stopContainerBackoffMax, stopContainerBackoffJitter, stopContainerBackoffMultiplier)
+	for i := 0; i < stopContainerMaxRetryCount; i++ {
+		md = engine.client.StopContainer(engine.ctx, dockerID, apiTimeoutStopContainer)
+		if md.Error == nil || md.Error.ErrorName() != dockerapi.DockerTimeoutErrorName {
+			return md
+		}
+		if i < stopContainerMaxRetryCount-1 {
+			time.Sleep(backoff.Duration())
+		}
+	}
+	if md.Error != nil && md.Error.ErrorName() == dockerapi.DockerTimeoutErrorName {
+		seelog.Warnf("Unable to stop container (%s) after %d attempts that timed out; giving up and marking container as stopped anyways", containerName, stopContainerMaxRetryCount)
```

Comment:
I do think it's the right choice to progress the task state, but I'd ask that we build out a mechanism to make this type of failure more discoverable, (not necessarily as part of the current PR).

While I agree that this change will unblock some, I fear that if we aren't being noisier about this we'll end up with instances that have abandoned containers lying around consuming resources.  And do we know whether a docker daemon that's caught in a deadlock will be fine servicing future tasks?